Documentation update regarding newly supported device ODROID M1S (#31540)
home-assistant/home-assistant.github.io: source/_includes/common-tasks/flashing_m1s_otg.md

## Flashing an ODROID-M1S

Home Assistant can be flashed to an ODROID-M1S by connecting the device directly to your computer via the USB-OTG connection on the front of the board. 

> Unlike other ODROID boards the M1S does not have a socket for an optional eMMC module. It also does not have a separate flash chip that holds a dedicated bootloader.

Reviewer: ```suggestion
Unlike other ODROID boards, the M1S does not have a socket for an optional  <abbr title="embedded MultiMediaCard">eMMC</abbr> module. It also does not have a separate flash chip that holds a dedicated bootloader.
```

> Instead the M1S has a build-in 64GB eMMC soldered directly on the board that holds a bootloader by default. This guide will show you how to flash HAOS to the build-in eMMC.

Reviewer: ```suggestion
Instead, the M1S has a build-in 64GB <abbr title="embedded MultiMediaCard">eMMC</abbr> soldered directly on the board that holds a bootloader by default. This guide will show you how to install the Home Assistant Operating System to the built-in <abbr title="embedded MultiMediaCard">eMMC</abbr>.
```

> Instead the M1S has a build-in 64GB eMMC soldered directly on the board that holds a petitboot bootloader by default. This guide will show you how to flash HAOS to the build-in eMMC.

Reviewer 2: ```suggestion
Instead the M1S has a build-in 64GB eMMC soldered directly on the board that holds a bootloader by default. This guide will show you how to flash HAOS to the build-in eMMC.
```

There's no petitboot - basically it's ARM firmware, followed by U-Boot SPL and U-Boot proper, but we can spare the reader of the gory details, just drop the "petitboot".

Author: Thank you for pointing that out. @darkxst already suggested to remove explicit "petitboot" references. That one here just slipped through...

> <ins>**Warning:</ins> Installing HAOS replaces the firmware and SPL on the eMMC with the mainline version provided by HAOS. As a result, it is not possible to use the SD card with the EMMC2UMS image anymore, because the mainline SPL is not compatible with U-Boot in the EMMC2UMS image at this time (February 2024). This does not pose any problem for standard use, just makes it more complicated in case you want to return to the Hardkernel-provided OS (see HK Recovery).**

Reviewer: ```suggestion
<ins>**Warning:</ins> Installing Home Assistant OS replaces the firmware and <abbr title="secondary program loader">SPL</abbr> on the <abbr title="embedded MultiMediaCard">eMMC</abbr> with the mainline version provided by the Home Assistant OS. As a result, it is not possible to use the SD card with the EMMC2UMS image anymore, because the mainline <abbr title="secondary program loader">SPL</abbr> is not compatible with U-Boot in the EMMC2UMS image at this time (February 2024). This does not pose any problem for standard use, just makes it more complicated in case you want to return to the Hardkernel-provided OS (see HK Recovery).**
```

### What you will need

> To flash your eMMC using OTG-USB, you will need the following items:

Reviewer: ```suggestion
To flash your <abbr title="embedded MultiMediaCard">eMMC</abbr> using <abbr title="USB-On-The-Go">USB-OTG</abbr>, you will need the following items:
```

- A small SD card
- Another computer
- USB 2.0 to micro-USB cable
- the special Hardkernel eMMC-to-USB-mass-storage image

> ### Boot into mass-storage mode

Reviewer: ```suggestion
### Boot into mass-storage mode

```

(These steps are identical to the official [Hardkernel wiki](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide?redirect=1#install_over_usb_from_pc) page.)

> 1. Download [ODROID-M1S_EMMC2UMS.img](https://dn.odroid.com/RK3566/ODROID-M1S/Installer/ODROID-M1S_EMMC2UMS.img)

Reviewer: ```suggestion
1. Download [ODROID-M1S_EMMC2UMS.img](https://dn.odroid.com/RK3566/ODROID-M1S/Installer/ODROID-M1S_EMMC2UMS.img).
```

2. Use [balena Etcher](https://www.balena.io/etcher/) or another tool to flash the UMS utility onto an SD card.

>    - use **Flash from file**. **Flash from URL** does not work on all systems.

Reviewer: ```suggestion
   - Use **Flash from file**. **Flash from URL** does not work on all systems.
```

>       (balena Etcher will complain that something went wring during flashin. You can ignore this message)

Reviewer: ```suggestion
      (balena Etcher will complain that something went wrong during flashing. You can ignore this message)
```

3. Plug-in that SD card to your ODROID-M1S and boot it.

> ### Flashing Home Assistant M1S

Reviewer: ```suggestion
### Flashing Home Assistant M1S

```

1. Download the latest stable version of Home Assistant OS for the [ODROID-M1S](https://github.com/home-assistant/operating-system/releases/download/{{site.data.version_data.hassos['odroid-m1s']}}/haos_odroid-m1s-{{site.data.version_data.hassos['odroid-m1s']}}.img.xz).

> 2. Apart from the HAOS image to flash (M1S instead of N2+ version), you can follow the N2+ step-by-step flashing guide [HERE](/common-tasks/os/#flashing-home-assistant).

Reviewer 3: Just refer them back to the generic odroid intructions rather than N2 guide?

Author: I think I don't agree here, because the reference at 2. is explicitly about the next steps how to flash the (internal) eMMC. Sure, a reference to the generic ODROID instructions are not entirely "wrong", but I think the reference to the really next steps makes more sense.

> #### _HK Recovery_

Reviewer: ```suggestion
#### _HK Recovery_

```

> If you want to restore your M1S back into Hardkernels initial state, you will have to restore the HK's bootloader.

Reviewer 2: ```suggestion
If you want to restore your M1S back into Hardkernel's initial state, you will have to restore the HK's bootloader.
```

> A reliable way of reflashing the eMMC in this case is to use HAOS booted from an SD card. To do that, insert the SD card with HAOS to the micro SD slot and plug the board in. Once the device boots to the HA CLI, enter `login` to enter the root shell and use `curl` to download an image and `dd` it to the eMMC block device:

Reviewer: ```suggestion
A reliable way of reflashing the eMMC in this case is to use Home Assistant OS booted from an SD card. To do that, insert the SD card with Home Assistant OS to the micro SD slot and plug the board in. Once the device boots to the Home Assistant CLI (command line), enter `login` to enter the root shell and use `curl` to download an image and `dd` it to the eMMC block device:
```

```sh
curl https://dn.odroid.com/RK3566/ODROID-M1S/Installer/ODROID-M1S_EMMC2UMS.img | dd of=/dev/mmcblk0
```

> This way the device will start in the UMS mode on the next boot with the SD card removed. Alternatively you can use the [Hardkernel installer image](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide#user_installer) directly instead of the EMMC2UMS image.

Reviewer: ```suggestion
This way, the device will start in the UMS mode on the next boot with the SD card removed. Alternatively, you can use the [Hardkernel installer image](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide#user_installer) directly instead of the EMMC2UMS image.
```

> This way the device will start in the UMS mode on the next boot with the SD card removed. Alternatively you can use the [Hardkernel installer image](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide#user_installer) directly instead of the EMMC2UMS image.

Reviewer 4: I don't think this should be in the user documentation. While convenient during development and for those know what they do, for lots of users this is probably confusing. We have it in the more power-user/developer oriented documentation in the OS repo (https://github.com/home-assistant/operating-system/blob/e60a2e4b68c061b920347b9394e5b95696d306cd/Documentation/boards/hardkernel/odroid-m1s.md#emmc).

Author: Thank you for reviewing that.

> I don't think this should be in the user documentation. While convenient during development and for those know what they do, for lots of users this is probably confusing. We have it in the more power-user/developer oriented documentation in the OS repo (https://github.com/home-assistant/operating-system/blob/e60a2e4b68c061b920347b9394e5b95696d306cd/Documentation/boards/hardkernel/odroid-m1s.md#emmc).

I personally am fine with keeping it there, because even as a less experienced user it would make me feel just comfortable if I directly see, that there is a "way back" from what I have been warned before. Or the other way: if I get warned about something and I don't see how/if I could revert that, this would make me kind of nervous.
If course I can remove it again and maybe just refer to the power-user guide if you or the majority asks me to do so. But my guts are fine with keeping it there ;) .

Reviewer 4: Sorry, my comment was a bit rushed and unclear.

> [...] it would make me feel just comfortable if I directly see, that there is a "way back" from what I have been warned before.

I agree with that.

However, the method described here is somewhat advanced. Folks will type this command into the web terminal, where access to  `/dev/mmcblk0` is not possible. The URL might become invalid. Every documentation needs to be maintained, if users are unclear issues being handled etc.  Sometimes more is more, but sometimes less is more :sweat_smile: 

There is a more generic way of recovering, which doesn't even need terminal access. and we should advice to use this by default. So I'd suggest to just document the Hardkernel installer image method, and link to the external wiki in this user documentation.

Reviewer 3: > There is a more generic way of recovering, which doesn't even need terminal access. and we should advice to use this by default. So I'd suggest to just document the Hardkernel installer image method, and link to the external wiki in this user documentation.

The Hardkernel images do not boot with mainline u-boot installed on eMMC though. That was the entire reason for the workaround in the first place. The only other option is shorting out maskROM pads to boot hardkernel images.

Reviewer 4: @darkxst so then this sentence is wrong?

> Alternatively you can use the [Hardkernel installer image](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide#user_installer) directly instead of the EMMC2UMS image.

Reviewer 4: Ah I understand, it is probably meant to be flashed using dd directly, similarly to ODROID-M1S_EMMC2UMS.img.

A few problems though: The command would be different as that image is xz compresseed. And this is definitely is not safe/guarnateed to work, as the image is larger and would overwrite the squashfs rootfs partition of HAOS. :cry:

Reviewer 4: Oh, nevermind, the commands are meant to be run from a HAOS running on a SD card. Yeah ok, then both methods should work and are safe.

The Hardkernel installer image can't be installed directly though because HAOS is missing unxz :cry: 

Also, I'd say that the command should work safely when booted from eMMC. The image is only 3MB, so this does not alter any of the mounted partitions. Having two HAOS installation on separate partition usually ends up in a bit chaotic situations since we rely on static partition UUID to find rootfs partitions etc.

I'll make a suggestion against the updated PR.

Reviewer 3: > Ah I understand, it is probably meant to be flashed using dd directly, similarly to ODROID-M1S_EMMC2UMS.img.

yes that is it, sairon added that comment to the docs, but guess he didnt test it

Reviewer 3: > The only other option is shorting out maskROM pads to boot hardkernel images.

One other option is to make our own UMS tool with mainline uboot. It is just uboot with UMS enabled in that Hardkernel image.

Author: I added the suggestion of @agners (see https://github.com/home-assistant/home-assistant.io/pull/31540/commits/86671fc6a59eac5d2e8fde9e6a9383b85df4fb25).
Can I clase this one as resolved?

Reviewer 2: @darkxst 
> The only other option is shorting out maskROM pads to boot hardkernel images.

But IIUC that should get you to the Rockchip download mode, which is even more involved procedure :grimacing:

> yes that is it, sairon added that comment to the docs, but guess he didnt test it

Yes, indeed I did not :sweat_smile: A simple piping through `xz -dc` should solve it, I guess I should be more explicit in the OS docs.

Reviewer 3: > But IIUC that should get you to the Rockchip download mode, which is even more involved procedure 😬

Only if it can't boot from SD card, will it fallback to RockUSB.

Reviewer 2: Ah, you're right. I got the impression it doesn't boot from the SD from [this post](https://forum.odroid.com/viewtopic.php?p=377662&sid=3fc9d7429603736e8dd7c2ca4c1ff904#p377662) but I just tested it behaves the way you say :+1:

1. Download [ODROID-M1S_EMMC2UMS.img](https://dn.odroid.com/RK3566/ODROID-M1S/Installer/ODROID-M1S_EMMC2UMS.img).
2. Use [balena Etcher](https://www.balena.io/etcher/) or another tool to flash the UMS utility onto an SD card.
   - Use **Flash from file**. **Flash from URL** does not work on all systems.
      (balena Etcher will complain that something went wrong during flashing. You can ignore this message)
3. Plug-in that SD card to your ODROID-M1S and boot it.

### Flashing Home Assistant M1S

1. Download the latest stable version of Home Assistant OS for the [ODROID-M1S](https://github.com/home-assistant/operating-system/releases/download/{{site.data.version_data.hassos['odroid-m1s']}}/haos_odroid-m1s-{{site.data.version_data.hassos['odroid-m1s']}}.img.xz).
2. Apart from the HAOS image to flash (M1S instead of N2+ version), you can follow the N2+ step-by-step flashing guide [HERE](/common-tasks/os/#flashing-home-assistant).

#### _HK Recovery_

If you want to restore your M1S back into Hardkernel's initial state, you will have to restore the HK's bootloader.

> A reliable way of reflashing the eMMC in this case is to use Home Assistant OS booted from an SD card. To do that, insert the SD card with Home Assistant OS to the micro SD slot and plug the board in. Once the device boots to the Home Assistant CLI (command line), enter `login` to enter the root shell and use `curl` to download an image and `dd` it to the eMMC block device:
> 
> ```sh
> curl https://dn.odroid.com/RK3566/ODROID-M1S/Installer/ODROID-M1S_EMMC2UMS.img | dd of=/dev/mmcblk0
> ```
> 
> This way, the device will start in the UMS mode on the next boot with the SD card removed. Alternatively, you can use the [Hardkernel installer image](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide#user_installer) directly instead of the EMMC2UMS image.

Reviewer 4: ```suggestion
A reliable way of reflashing the eMMC with an operating system of youur choice is to use Home Assistant OS to flash the EMMC2UMS image which turns the ODROID-M1S into USB Mass Storage device. Once you have flashed the EMMC2UMS image, you can flash any OS again. You will need a micro USB cable to connect ODROID-M1S to PC.

Note: This commands will render your current Home Assistant OS installation unbootable!

Use the local terminal (HDMI/keyboard) to access the system console. On the Home Assistant CLI (command line), enter `login` to enter the root shell and use `curl` to download an image and `dd` it to the eMMC block device:

```sh
curl https://dn.odroid.com/RK3566/ODROID-M1S/Installer/ODROID-M1S_EMMC2UMS.img | dd of=/dev/mmcblk0
```

This way, the device will start in the UMS mode on the next boot with the SD card removed. Follow the [Install over USB from PC](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide#install_over_usb_from_pc) to install a different operating system.
```

Author: corrected a typo and added the last part that you seem to have lost (or just weird formatted in the suggestion post ;) )

Reviewer 4: @Jojo-A uh whoops, I see thanks.

However, it seems the link got lost. Can you make sure that we link to https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide#install_over_usb_from_pc in that last sentence?

Author: Oh, whoops!
Done :) .

## Flashing an ODROID-M1S

Home Assistant can be flashed to an ODROID-M1S by connecting the device directly to your computer via the USB-OTG connection on the front of the board. 

> Unlike other ODROID boards the M1S does not have a socket for an optionl eMMC module. It also does not have a separate flash chip that holds a dedicated bootloader.

Reviewer 3: typo, "optionl"

Author: Done

> Instead the M1S has a build-in 64GB eMMC soldered directly on the board that holds a petitboot bootloader by default. This guide will show you how to flash HAOS to the build-in eMMC.

Reviewer 3: Please see the updates to the OS docs here, and reword in line with that and drop references to Petitboot: 
https://github.com/home-assistant/operating-system/blob/dev/Documentation/boards/hardkernel/odroid-m1s.md

It also includes a new recovery method that doesnt require using maskrom pin. Maybe you could add the Recovery steps as its own section?

Author: Done. Updated the initial section and added a separate HK Recovery section

If you wish to return to any official Hardkernel images then short the maskRom pads near the 40pin connector while booting from a [Hardkernel recovery image](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide?redirect=1#user_installer).**

### What you will need

To flash your eMMC using OTG-USB, you will need the following items:

- A small SD card
- Another computer
- USB 2.0 to micro-USB cable
- the special Hardkernel eMMC-to-USB-mass-storage image

### Boot into mass-storage mode
(These steps are identical to the official [Hardkernel wiki](https://wiki.odroid.com/odroid-m1s/getting_started/os_installation_guide?redirect=1#install_over_usb_from_pc) page.)

1. Download [ODROID-M1S_EMMC2UMS.img](https://dn.odroid.com/RK3566/ODROID-M1S/Installer/ODROID-M1S_EMMC2UMS.img)
2. Use [balena Etcher](https://www.balena.io/etcher/) or another tool to flash the UMS utility onto an SD card.
   - use **Flash from file**. **Flash from URL** does not work on all systems.
      (balena Etcher will complain that something went wring during flashin. You can ignore this message)
3. Plug-in that SD card to your ODROID-M1S and boot it.

> ### Flashing Home Assistant

Reviewer 3: not sure you need to duplicate the steps from the main installation page?

Author: Good point. True, there is nothing wrong in duplicating the description. But maybe just a link/reference to the flashing guide of the N2+ could be sufficient. I try to implement it to see how it looks like...

Author: Okay, done. I hope the link to the N2+'s flashing guide works in the final version. Syntax should be fine from what I have read...

1. Connect the ODROID-M1S to your PC via the micro-USB port. 
2. When the ODROID-M1S is recognized as a USB connected storage device, you can flash the eMMC with [Etcher](https://www.balena.io/etcher/).
![ums](https://wiki.odroid.com/_media/odroid-m1s/getting_started/disk_management.png)
   - Use the latest stable version of Home Assistant OS for the [ODROID-M1S](https://github.com/home-assistant/operating-system/releases/download/{{site.data.version_data.hassos['odroid-m1s']}}/haos_odroid-m1s-{{site.data.version_data.hassos['odroid-m1s']}}.img.xz).
   - In Balena, use **Flash from file**. **Flash from URL** does not work on all systems.

3. When the flash process is complete, disconnect the ODROID-M1S from your PC.
   - Remove the power cable.

>    - Remove the micro-USB cabel

Reviewer 3: typo

Author: Done
